### ISSUE TITLE
Provide support for both AMD and NVIDIA GPU : consider HIP ?

### ISSUE THREAD

**Issue by Firefly35** [colmap/colmap#1173]
CUDA is a nvidia only acceleration solution, provide a mean to support acceleration on AMD GPUs

For instance, consider migrating the CUDA code using https://github.com/ROCm-Developer-Tools/HIP. It would both provide support for AMD and NVIDIA GPUs.

**Comment by whuaegeanse** [comment]
This may be a huge job, and related algorithms such as feature extraction, feature matching, and MVS need to be rewritten.

**Comment by Firefly35** [comment]
I thought after reading HIP presentation that it was made to ease the conversion from CUDA to HIP process.
Tools are even provided : https://github.com/ROCm-Developer-Tools/HIPIFY/blob/master/README.md
Did you already get a deep look to it ?

**Comment by granger35** [comment]
+1

**Comment by revisionarian** [comment]
HIP programs only run on a small fraction of AMD GPUs.

[COLMAP-CL](https://github.com/openphotogrammetry/colmap-cl) is a version of COLMAP that is accelerated on any GPU that supports OpenCL. This includes most AMD, NVIDIA, Intel, ARM, etc. GPUs. The Windows version of COLMAP-CL is available here: https://github.com/openphotogrammetry/colmap-cl.

**Comment by ahojnnes** [comment]
@revisionarian That's very cool. I looked at the repository but didn't see any source code. Do you plan to release it?

**Comment by revisionarian** [comment]
> @revisionarian That's very cool. I looked at the repository but didn't see any source code. Do you plan to release it?

Yes. Would you be interested in a pull request to your main COLMAP repository? The CUDA vs. OpenCL support could be selectable at either compile-time or runtime.
